clients/filesystem-hadoop/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java

```diff
@@ -0,0 +1,404 @@
+  private Path workingDirectory;
+  private URI uri;
+  private GravitinoClient client;
+  private String metalakeName;
+  private Cache<NameIdentifier, Pair<Fileset, FileSystem>> authedFilesetCache;
+  private ScheduledThreadPoolExecutor scheduler;
+
+  @Override
+  public void initialize(URI name, Configuration configuration) throws IOException {
+    if (!name.toString().startsWith(GravitinoVirtualFileSystemConfiguration.GVFS_FILESET_PREFIX)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Unsupported file system scheme: %s for %s.",
+              name.getScheme(), GravitinoVirtualFileSystemConfiguration.GVFS_SCHEME));
+    }
+
+    int maxCapacity =
+        configuration.getInt(
+            GravitinoVirtualFileSystemConfiguration.FS_GRAVITINO_FILESET_CACHE_MAX_CAPACITY_KEY,
+            GravitinoVirtualFileSystemConfiguration
+                .FS_GRAVITINO_FILESET_CACHE_MAX_CAPACITY_DEFAULT);
+    Preconditions.checkArgument(maxCapacity > 0, "Cache max capacity should be greater than 0");
+
+    long evictionMillsAfterAccess =
+        configuration.getLong(
+            GravitinoVirtualFileSystemConfiguration
+                .FS_GRAVITINO_FILESET_CACHE_EVICTION_MILLS_AFTER_ACCESS_KEY,
+            GravitinoVirtualFileSystemConfiguration
+                .FS_GRAVITINO_FILESET_CACHE_EVICTION_MILLS_AFTER_ACCESS_DEFAULT);
+    Preconditions.checkArgument(
+        evictionMillsAfterAccess > 0, "Cache eviction mills after access should be greater than 0");
+
+    initializeCache(maxCapacity, evictionMillsAfterAccess);
+
+    // initialize the Gravitino client
+    String serverUri =
+        configuration.get(GravitinoVirtualFileSystemConfiguration.FS_GRAVITINO_SERVER_URI_KEY);
+    Preconditions.checkArgument(
+        StringUtils.isNotBlank(serverUri), "Gravitino server uri is not set in the configuration");
+
+    this.metalakeName =
+        configuration.get(GravitinoVirtualFileSystemConfiguration.FS_GRAVITINO_CLIENT_METALAKE_KEY);
+    Preconditions.checkArgument(
+        StringUtils.isNotBlank(metalakeName), "Gravitino metalake is not set in the configuration");
+
+    // TODO Need support more authentication types, now we only support simple auth
+    this.client =
+        GravitinoClient.builder(serverUri).withMetalake(metalakeName).withSimpleAuth().build();
+
+    this.workingDirectory = new Path(name);
+    this.uri = URI.create(name.getScheme() + "://" + name.getAuthority());
+
+    setConf(configuration);
+    super.initialize(uri, getConf());
+  }
+
+  private void initializeCache(int maxCapacity, long expireAfterAccess) {
+    // Since Caffeine does not ensure that removalListener will be involved after expiration
+    // We use a scheduler with one thread to clean up expired clients.
+    this.scheduler = new ScheduledThreadPoolExecutor(1, newDaemonThreadFactory());
+
+    this.authedFilesetCache =
```

Comment:
What is the meaning of `authedFilesetCache`, I cannot get the meaning of "auth" here. Can you please explain more?

Reply:
The Fileset in the Cache here is the Fileset loaded through the `GravitinoClient` with the `user Token` request from the `Server`. If the Fileset is not authenticated, it will not be cached. But currently, gvfs only supports `SimpleAuth`. It's ok to not stress `auth`.

Reply:
Yep, I think we don't have emphasize the "auth" here, just `filesetCache` should be enough.